src/cmd/binary.js

```diff
@@ -25,47 +25,39 @@ License along with this program; if not, see <http://www.gnu.org/licenses/>.
  ******************************************************************************
  */
 
-const when = require('when');
 const fs = require('fs');
 const path = require('path');
+const VError = require('verror');
 const chalk = require('chalk');
 const Parser = require('binary-version-reader').HalModuleParser;
 const utilities = require('../lib/utilities.js');
+const ensureError = utilities.ensureError;
 
 class BinaryCommand {
-	constructor(options) {
-		this.options = options;
-	}
-
-	inspectBinary() {
-		const binaryFile = this.options.params.filename;
-
-		if (!binaryFile || !fs.existsSync(binaryFile)) {
-			console.error('Please specify a binary file');
-			return when.reject();
-		}
-		const dfd = when.defer();
-		const parser = new Parser();
-		parser.parseFile(binaryFile, (fileInfo, err) => {
-			if (err) {
-				console.error(err);
-				return dfd.reject();
+	inspectBinary(binaryFile) {
+		return new Promise((fulfill, reject) => {
+			if (!fs.existsSync(binaryFile)) {
+				throw new VError(`Binary file not found ${binaryFile}`);
 			}
+			const parser = new Parser();
+			parser.parseFile(binaryFile, (fileInfo, err) => {
+				if (err) {
+					return reject(new VError(ensureError(err), `Could not parse ${binaryFile}`));
```

Comment:
_trying to understand JS better_

On line 40, the error is only thrown. Is this on purpose? If so, is there a difference between reject and throw in this case?

Reply:
No, line 40 is a bug.

Reply:
In general I prefer to throw errors in both async and sync code so I don't have to use 2 patterns for what feels like the same thing. The challenge comes in when mixing callback-based code with Promise based code. It can be easy to make a mistake.